### PR TITLE
Add unobtrusive method tracing

### PR DESCRIPTION
**Pull request by norbertnytko**
Follow up to issue https://github.com/DataDog/dd-trace-rb/issues/1197

### CODE REVIEW

File: lib/ddtrace/method_tracer.rb [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+      def trace_singleton_methods(klass, *method_names)
+        method_names.each { |method_name| instrument_singleton_method(klass, method_name) }
+        prepend_container(klass)
+      end
+
+      private
+
+      def instrument_method(klass, method)
+        Container.define_method(method) do |*args, &block|
```
Comment (norbertnytko) [inline lib/ddtrace/method_tracer.rb:68]: We can actually skip this step if method is defined in `Container`

File: lib/ddtrace/method_tracer.rb [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+          Datadog.tracer.trace(OPERATION_NAME, options) { super(*args, &block) }
+        end
+      end
+
+      def prepend_container(klass)
+        klass.prepend Container
+      end
+
+      # Not so private, Container holds all dynamically created methods
+      module Container
```
Comment (norbertnytko) [inline lib/ddtrace/method_tracer.rb:102]: Unfortunate couldn't figure out way to use anonymous module here

**Comment by norbertnytko** [comment]
@marcotc

**Comment by ericmustin** [comment]
Sorry for the delay here in reviewing. Generally I think this is a good approach and we appreciate the contribution, and if i'm understanding the implementation correctly, we're successfully avoiding the pitfall [mentioned here](https://github.com/DataDog/dd-trace-rb/pull/670#discussion_r388595271) by only prepending the `Container` module once https://github.com/DataDog/dd-trace-rb/pull/1215/files#diff-ccf38a2ec0604d3bdf40d8ed352af6fca58d18fcf29578f82c5b0c5155b4a864R58-R59. It would be good to get a second set of eyes on this I will try to have one of my colleagues take a look as well.

File: lib/ddtrace/method_tracer.rb [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+      def trace_singleton_methods(klass, *method_names)
+        method_names.each { |method_name| instrument_singleton_method(klass, method_name) }
+        prepend_container(klass)
+      end
+
+      private
+
+      def instrument_method(klass, method)
+        Container.create_method(method) do |*args, &block|
```
Comment (marcotc) [inline lib/ddtrace/method_tracer.rb:70]: Wouldn't `Container` here always refer to the same `Datadog::MethodTracer::Container` module? Meaning `create_method` would be creating a method on a globally shared singleton `Container`?

If this assumption of mine is correct, and `Instrumentation#trace_methods` and `Instrumentation#trace_singleton_methods` have to create new modules for each `klass.send(:prepend, Container)` call, I think that we should ensure that `#trace_methods` and `#trace_singleton_methods` reuse the prepended module if it's already there.

File: lib/ddtrace/method_tracer.rb [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+      end
+
+      # Not so private, Container holds all dynamically created methods
+      module Container
+        # This trick is required in order to work with singleton methods
+        module ClassMethods
+          # Unfortunately is private for ruby 2.4 and below
+          def self.create_method(name, &block)
+            define_method(name, &block)
+          end
```
Comment (marcotc) [inline lib/ddtrace/method_tracer.rb:108]: If the goal is to expose a private method, wouldn't it be possible to instead call `send(:define_method, ...)` at the at the call site, thus avoiding this method definition?

**Review by marcotc — COMMENTED** [review]
Sorry for the delay on the feedback, @norbertnytko.

We've been wanting this feature for a while, but it introduces a new and very opinionated API (and for good reason), so we were being careful not to rush a review here.

I think that the interface and implementation look very good, so I don't see major design changes required to get it merged.

I left a few comments, mostly about trying to ensure the correctness of the implementation. Please let me know what you think!

File: lib/ddtrace/method_tracer.rb (hunk 4 above)
Comment (norbertnytko) [reply lib/ddtrace/method_tracer.rb]: Changed 👌

**Comment by marcotc** [comment]
Thank you so much for the updates, @norbertnytko. I'll take another thorough look at the new changes soon.

File: lib/ddtrace/method_tracer.rb (hunk 3 above)
Comment (norbertnytko) [reply lib/ddtrace/method_tracer.rb]: Your assumption is correct [added condition](https://github.com/DataDog/dd-trace-rb/pull/1215/files#diff-ccf38a2ec0604d3bdf40d8ed352af6fca58d18fcf29578f82c5b0c5155b4a864R70)

File: lib/ddtrace/method_tracer.rb [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    def trace_methods(klass, *method_names)
+      instrumentation.trace_methods klass, *method_names
+    end
+
+    def trace_singleton_methods(klass, *method_names)
+      instrumentation.trace_singleton_methods klass, *method_names
+    end
+
+    def instrumentation
+      @_instrumentation ||= Instrumentation.new
```
Comment (marcotc) [inline lib/ddtrace/method_tracer.rb:23]: It's good to give a hint to users that this environment variable present in their application comes from Datadog:
```suggestion
      @dd_instrumentation ||= Instrumentation.new
```

**Review by marcotc — COMMENTED** [review]
One super minor comment and we are ready to merge it. 👍

**Comment by marcotc** [comment]
@norbertnytko thank you so much for this PR! We think that this is the most polished and clean implementation of a tracer mixin to date 🎉.

There are finishing touches we'd like to have shipped with the tracer mixin (e.g. documentation, a way to provide options to the created span, review module/method naming convention), but most of these require close coordination with the internal team.

At this time, because our internal priorities are in alignment with the work you've done here, we're taking the responsibility of addressing these last changes and getting your work merged.

I'm continuing this work in another PR here: #1348. I would love to hear your feedback when it's all finally done, to ensure that we are addressing the exact use case you have.

Again, thank you so much for all the help! 🙇

**Comment by norbertnytko** [comment]
😢

**Pull request closed without merging**